be/tx_bulk.c

```diff
@@ -52,52 +87,158 @@ enum {
+	int                       rc;
+
+	M0_PRE(M0_IS0(tb));
+	M0_PRE(tb_cfg->tbc_partitions_nr > 0);
+	/* Can't have more partitions that workers. For now. */
+	M0_PRE(tb_cfg->tbc_partitions_nr <= tb_cfg->tbc_workers_nr);
+	M0_PRE(tb_cfg->tbc_work_items_per_tx_max > 0);
+	M0_PRE(tb_cfg->tbc_q_cfg.bqc_item_length == 0);
 
 	tb->btb_cfg = *tb_cfg;
-	m0_be_domain__group_limits(tb->btb_cfg.tbc_dom,
-				   &group_nr, &tx_per_group);
-	tb->btb_worker_nr = be_tx_bulk_worker_nr(tb, group_nr, tx_per_group);
-	tb->btb_done_nr = 0;
-	tb->btb_stopping = false;
+	tb->btb_tx_open_failed = false;
 	tb->btb_done = false;
+	tb->btb_termination_in_progress = false;
 	m0_mutex_init(&tb->btb_lock);
-	M0_ALLOC_ARR(tb->btb_worker, tb->btb_worker_nr);
+	M0_ALLOC_ARR(tb->btb_worker, tb->btb_cfg.tbc_workers_nr);
 	rc = tb->btb_worker == NULL ? -ENOMEM : 0;
-	if (rc != 0)
+	if (rc != 0) {
+		m0_mutex_fini(&tb->btb_lock);
+		return M0_ERR(rc);
+	}
+	/* XXX error handling */
+	M0_ALLOC_ARR(tb->btb_q, tb->btb_cfg.tbc_partitions_nr);
+	M0_ASSERT(tb->btb_q != NULL);
+	tb->btb_cfg.tbc_q_cfg.bqc_item_length = sizeof(struct be_tx_bulk_item);
+	for (i = 0; i < tb_cfg->tbc_partitions_nr; ++i) {
+		rc = m0_be_queue_init(&tb->btb_q[i], &tb->btb_cfg.tbc_q_cfg);
+		if (rc != 0)
+			break;
+	}
+	if (rc != 0) {
+		m0_free(tb->btb_worker);
 		m0_mutex_fini(&tb->btb_lock);
+		return M0_ERR(rc);
+	}
+	m0_be_op_init(&tb->btb_kill_put_op);
+	/* XXX take it from elsewhere */
+	localities_nr = m0_fom_dom()->fd_localities_nr;
+	for (i = 0; i < tb->btb_cfg.tbc_workers_nr; ++i) {
+		worker = &tb->btb_worker[i];
+		worker_partition = i % tb->btb_cfg.tbc_partitions_nr;
+		/*
+		 * Each locality has a partitition assigned, and each worker
+		 * will be in one of such partititons.
+		 */
+		worker_locality  = worker_partition % localities_nr;
+		*worker = (struct be_tx_bulk_worker){
+			.tbw_index                = i,
+			.tbw_tb                   = tb,
+			.tbw_items_nr             = 0,
+			.tbw_queue_get_successful = false,
+			.tbw_grp                  =
+				m0_locality_get(worker_locality)->lo_grp,
+			.tbw_rc                   = 0,
+			.tbw_queue_get            = {
+				.sa_cb    = &be_tx_bulk_queue_get_cb,
+				.sa_datum = worker,
+			},
+			.tbw_init                 = {
+				.sa_cb    = &be_tx_bulk_init_cb,
+				.sa_datum = worker,
+			},
+			.tbw_close                = {
+				.sa_cb    = &be_tx_bulk_close_cb,
+				.sa_datum = worker,
+			},
+			.tbw_finish               = {
+				.sa_cb    = &be_tx_bulk_finish_cb,
+				.sa_datum = worker,
+			},
+			.tbw_failed               = false,
+			.tbw_done                 = false,
+			.tbw_partition            = worker_partition,
+		};
+		m0_be_op_init(&worker->tbw_op);
+		m0_be_op_callback_set(&worker->tbw_op,
+		                      &be_tx_bulk_queue_get_done_cb,
+		                      worker, M0_BOS_DONE);
+		M0_ALLOC_ARR(worker->tbw_item,
+			     tb->btb_cfg.tbc_work_items_per_tx_max);
+		M0_ASSERT(worker->tbw_item != NULL);  /* XXX */
```

Comment:
Please replace XXX with proper action item. So it can be taken up later.

Reply:
Added a comment.